suite/tests/api/ir_rv64_addr.c

```diff
@@ -0,0 +1,176 @@
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of ISCAS nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ISCAS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Define DR_FAST_IR to verify that everything compiles when we call the inline
+ * versions of these routines.
+ */
+#ifndef STANDALONE_DECODER
+#    define DR_FAST_IR 1
+#endif
+
+/* Uses the DR API, using DR as a standalone library, rather than
+ * being a client library working with DR on a target program.
+ */
+
+#include "configure.h"
+#include "dr_api.h"
+#include "dr_defines.h"
+#include "dr_ir_utils.h"
+#include "tools.h"
+
+#include "ir_riscv64.h"
+
+static byte *
+test_instr_encoding_auipc(void *dc, uint opcode, app_pc instr_pc, instr_t *instr)
+{
+    instr_t *decin;
+    byte *pc, *next_pc;
+
+    ASSERT(instr_get_opcode(instr) == opcode);
+    instr_disassemble(dc, instr, STDERR);
+    print("\n");
+    ASSERT(instr_is_encoding_possible(instr));
+    pc = instr_encode_to_copy(dc, instr, buf, instr_pc);
+    ASSERT(pc != NULL);
+    decin = instr_create(dc);
+    next_pc = decode_from_copy(dc, buf, instr_pc, decin);
+    ASSERT(next_pc != NULL);
+    if (!instr_same(instr, decin)) {
+        print("Disassembled as:\n");
+        instr_disassemble(dc, decin, STDERR);
+        print("\n");
+        ASSERT(instr_same(instr, decin));
+    }
+
+    instr_destroy(dc, instr);
+    instr_destroy(dc, decin);
+    return pc;
+}
+
+static void
+test_jump_and_branch(void *dc)
+{
+    byte *pc;
+    instr_t *instr;
+
+    instr = INSTR_CREATE_lui(dc, opnd_create_reg(DR_REG_A0),
+                             opnd_create_immed_int(42, OPSZ_20b));
+    pc = test_instr_encoding(dc, OP_lui, instr);
+
+    instr = INSTR_CREATE_auipc(dc, opnd_create_reg(DR_REG_A0),
+                               OPND_CREATE_ABSMEM(pc + (3 << 12), OPSZ_0));
+    test_instr_encoding_auipc(dc, OP_auipc, pc, instr);
```

Comment:
Not sure I understand this: `pc` is the PC *after* the `lui` above encoded at `buf`; so this is encoding `auipc` at `buf` as though it were really at `buf+4` or sthg?  Why pick that PC?  The original code at HEAD was encoding as though at `buf`: just 4 bytes prior?  I guess `buf` is hidden from here: but it was there in the original file; and see the other comment where we might want to pass it in.  Is it just that you want to abstract away `buf` from this code here?

Reply:
> Is it just that you want to abstract away `buf` from this code here?

Yes, it shouldn't be necessary. In the new commit, the address of `buf` is assigned to `pc` directly and the (possibly confusing) encoding test of `lui` is eliminated.